handler.go

```diff
@@ -0,0 +1,276 @@
+// custom headers on the response.
+type JSONHandler func(http.Header, Params) (interface{}, error)
+
+// ErrorHandler is like httprouter.Handle except it returns an error
+// which may be returned as the error body of the response.
+// An ErrorHandler function should not itself write to the ResponseWriter
+// if it returns an error.
+type ErrorHandler func(http.ResponseWriter, Params) error
+
+// HandleJSON returns a handler that writes the return value
+// of handle as a JSON response. If handle returns an error,
+// it is passed through the error mapper.
+func (e ErrorMapper) HandleJSON(handle JSONHandler) httprouter.Handle {
+	f := func(w http.ResponseWriter, p Params) error {
+		val, err := handle(w.Header(), p)
+		if err != nil {
+			return errgo.Mask(err, errgo.Any)
+		}
+		return WriteJSON(w, http.StatusOK, val)
+	}
+	return e.HandleErrors(f)
+}
+
+// HandleErrors returns a handler that passes
+// any non-nil error returned by handle through the
+// error mapper and writes it as a JSON response.
+func (e ErrorMapper) HandleErrors(handle ErrorHandler) httprouter.Handle {
+	return func(w http.ResponseWriter, req *http.Request, p httprouter.Params) {
+		w1 := responseWriter{
+			ResponseWriter: w,
+		}
+		if err := handle(&w1, Params{req, p}); err != nil {
+			// We write the error only if the header hasn't
+			// already been written, because if it has, then
+			// we will not be able to set the appropriate error
+			// response code, and there's a danger that we
+			// may be corrupting output by appending
+			// a JSON error message to it.
+			if !w1.headerWritten {
+				e.WriteError(w, err)
+			}
+			// TODO log the error?
+		}
+	}
+}
+
+// WriteError writes an error to a ResponseWriter
+// and sets the HTTP status code.
+func (e ErrorMapper) WriteError(w http.ResponseWriter, err error) {
+	status, resp := e(err)
+	WriteJSON(w, status, resp)
+}
+
+// WriteJSON writes the given value to the ResponseWriter
+// and sets the HTTP status to the given code.
+func WriteJSON(w http.ResponseWriter, code int, val interface{}) error {
+	// TODO consider marshalling directly to w using json.NewEncoder.
+	// pro: this will not require a full buffer allocation.
+	// con: if there's an error after the first write, it will be lost.
+	data, err := json.Marshal(val)
+	if err != nil {
+		// TODO(rog) log an error if this fails and lose the
+		// error return, because most callers will need
+		// to do that anyway.
+		return errgo.Mask(err)
+	}
+	w.Header().Set("content-type", "application/json")
+	w.WriteHeader(code)
+	w.Write(data)
+	return nil
+}
+
+// Ensure statically that responseWriter does implement http.Flusher.
+var _ http.Flusher = (*responseWriter)(nil)
+
+// responseWriter wraps http.ResponseWriter but allows us
+// to find out whether any body has already been written.
+type responseWriter struct {
+	headerWritten bool
+	http.ResponseWriter
+}
+
+func (w *responseWriter) Write(data []byte) (int, error) {
+	w.headerWritten = true
```

Comment:
Because writing data to the response writer implicitly writes the header (you can't
add a header after the data in HTTP)